examples/cms/src/FieldMapping.tsx

```diff
@@ -0,0 +1,225 @@
+                        onFieldNameChange(field.id, fieldConfig.name)
+                    } else {
+                        onFieldNameChange(field.id, e.target.value)
+                    }
+                }}
+            />
+        </Fragment>
+    )
+}
+
+export function FieldMapping({
+    savedFieldsConfig,
+    existingFields,
+    dataSource,
+    savedSlugFieldId,
+    onSubmit,
+}: {
+    savedFieldsConfig: FieldConfig[] | undefined
+    existingFields: ManagedCollectionField[]
+    dataSource: DataSource
+    savedSlugFieldId: string | null
+    onSubmit: (dataSource: DataSource, fields: FieldConfig[], slugFieldId: string) => Promise<void>
+}) {
+    const [fields, setFields] = useState<FieldConfig[]>(
+        savedFieldsConfig ?? computeFieldConfigs(existingFields, dataSource)
+    )
+    const [disabledFieldIds, setDisabledFieldIds] = useState<Set<string>>(
+        new Set(savedFieldsConfig?.filter(field => field.isNew).map(field => field.field!.id))
+    )
+
+    const possibleSlugFields = useMemo(
+        () =>
+            fields
+                .map(field => field.field)
+                .filter(isNotNull)
+                .filter(field => !disabledFieldIds.has(field.id) && field.type === "string"),
+        [fields, disabledFieldIds]
+    )
+    const [slugFieldId, setSlugFieldId] = useState<string | null>(savedSlugFieldId ?? possibleSlugFields[0]?.id ?? null)
+
+    const [isSyncing, setIsSyncing] = useState(false)
+
+    const handleFieldNameChange = (fieldId: string, name: string) => {
+        setFields(prev =>
+            prev.map(field => (field.field?.id === fieldId ? { ...field, field: { ...field.field, name } } : field))
+        )
+    }
+
+    const handleFieldToggle = (fieldId: string) => {
+        setDisabledFieldIds(current => {
+            const nextSet = new Set(current)
+            if (nextSet.has(fieldId)) {
+                nextSet.delete(fieldId)
+
+                // If we're re-enabling a string field and there's no valid slug field,
+                // set this field as the slug field
+                const field = fields.find(field => field.field?.id === fieldId)
+                if (field?.field?.type === "string") {
+                    const currentSlugField = fields.find(field => field.field?.id === slugFieldId)
+                    if (!currentSlugField || nextSet.has(slugFieldId ?? "")) {
+                        setSlugFieldId(fieldId)
+                    }
+                }
+            } else {
+                nextSet.add(fieldId)
+
+                // If the disabled field is the slug field, update it to the next
+                // possible slug field
+                if (fieldId === slugFieldId) {
+                    const nextSlugField = possibleSlugFields.find(field => field.id !== fieldId)
+                    if (nextSlugField) {
+                        setSlugFieldId(nextSlugField.id)
+                    }
+                }
+            }
```

Comment:
I have a feeling that this can be simplified a lot. But my brain is too fried right now

Reply:
I did a pass on this!